// @ngxs-labs/dispatch-decorator/esm2015/lib/core/internals/static-injector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgZone } from '@angular/core';
import { Store } from '@ngxs/store';
class NgxsDispatchPluginModuleNotImported extends Error {
    constructor() {
        super(...arguments);
        this.message = 'NgxsDispatchPluginModule is not imported';
    }
}
if (false) {
    /** @type {?} */
    NgxsDispatchPluginModuleNotImported.prototype.message;
}
/** @type {?} */
let injector = null;
/**
 * @template T
 * @param {?} actual
 * @return {?}
 */
function assertDefined(actual) {
    if (actual == null) {
        throw new NgxsDispatchPluginModuleNotImported();
    }
}
/**
 * @param {?} parentInjector
 * @return {?}
 */
export function setInjector(parentInjector) {
    injector = parentInjector;
}
/**
 * @return {?}
 */
export function getStore() {
    assertDefined(injector);
    return (/** @type {?} */ (injector)).get(Store);
}
/**
 * @return {?}
 */
export function getNgZone() {
    assertDefined(injector);
    return (/** @type {?} */ (injector)).get(NgZone);
}
